File: packages/android/src/main/java/com/kozmos/components/RouteSummary/RouteSummary.kt
package com.kozmos.components.routesummary

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Navigation
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import androidx.compose.material3.LocalContentColor
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import com.kozmos.components.button.KozmosButton
import com.kozmos.components.surface.KozmosSurfaceDefaults
import com.kozmos.components.surface.KozmosSurfaceStyle
import com.kozmos.components.button.KozmosButtonEmotion
import com.kozmos.components.button.KozmosButtonSize
import com.kozmos.components.button.KozmosButtonVariant
import com.kozmos.tokens.KozmosThemeTokens
import com.kozmos.tokens.KozmosDimensions

enum class KozmosRouteSummaryState {
    Active,
    Preview
}

@Composable
fun KozmosRouteSummary(
    etaText: String,
    distanceText: String,
    onEndRoute: () -> Unit,
    modifier: Modifier = Modifier,
    state: KozmosRouteSummaryState = KozmosRouteSummaryState.Active,
    onStartNavigation: (() -> Unit)? = null,
    surface: KozmosSurfaceStyle = KozmosSurfaceStyle.Solid,
    transportModeIcon: (@Composable () -> Unit)? = null
) {
    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusPanel),
        color = KozmosSurfaceDefaults.tint(surface),
        tonalElevation = 6.dp,
        shadowElevation = 12.dp,
        border = KozmosSurfaceDefaults.border(surface)
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(KozmosDimensions.primitivesLayoutSpacing200),
            verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing200)
        ) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.fillMaxWidth()
            ) {
                if (transportModeIcon != null) {
                    Surface(
                        modifier = Modifier.size(40.dp),
                        shape = CircleShape,
                        color = KozmosThemeTokens.primitivesColorsTheme500.copy(alpha = 0.12f)
                    ) {
                        androidx.compose.foundation.layout.Box(contentAlignment = Alignment.Center) {
                            transportModeIcon()
                        }
                    }
                    Spacer(modifier = Modifier.size(KozmosDimensions.primitivesLayoutSpacing150))
                }

                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        text = etaText,
                        style = MaterialTheme.typography.titleLarge,
                        color = KozmosThemeTokens.primitivesColorsForeground100
                    )
                    Text(
                        text = distanceText,
                        style = MaterialTheme.typography.bodyMedium,
                        color = KozmosThemeTokens.primitivesColorsForeground500
                    )
                }

                if (state == KozmosRouteSummaryState.Active) {
                    IconButton(
                        onClick = onEndRoute,
                        modifier = Modifier
                            .size(40.dp)
                            .semantics { contentDescription = "End route" }
                    ) {
                        Icon(
                            imageVector = Icons.Default.Close,
                            contentDescription = null,
                            tint = KozmosThemeTokens.primitivesColorsEmotionalDanger600
                        )
                    }
                }
            }

            if (state == KozmosRouteSummaryState.Preview && onStartNavigation != null) {
                KozmosButton(
                    onClick = onStartNavigation,
                    modifier = Modifier.fillMaxWidth()
                ) {
                    Icon(Icons.Default.Navigation, contentDescription = null)
                    Text("Start Navigation")
                }
            }
        }
    }
}

/**
 * The navigation layout: the destination's name with End beside it in the
 * danger outline; the time, distance and arrival on one row; the caller's
 * `progress` — a `KozmosRouteProgressRail`, in the products — below. The
 * layout above is unchanged.
 */
@Composable
fun KozmosRouteSummary(
    destination: String,
    durationText: String,
    distanceText: String,
    onEndRoute: () -> Unit,
    modifier: Modifier = Modifier,
    arrivalText: String? = null,
    endLabel: String = "End",
    surface: KozmosSurfaceStyle = KozmosSurfaceStyle.Solid,
    progress: (@Composable () -> Unit)? = null
) {
    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusPanel),
        color = KozmosSurfaceDefaults.tint(surface),
        tonalElevation = 6.dp,
        shadowElevation = 12.dp,
        border = KozmosSurfaceDefaults.border(surface)
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(KozmosDimensions.primitivesLayoutSpacing200),
            verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing150)
        ) {
            Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.fillMaxWidth()) {
                Text(
                    text = destination,
                    style = MaterialTheme.typography.titleLarge.copy(fontWeight = FontWeight.SemiBold),
                    color = KozmosThemeTokens.primitivesColorsForeground100,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f).semantics { heading() }
                )
                Spacer(modifier = Modifier.size(KozmosDimensions.primitivesLayoutSpacing150))
                KozmosButton(
                    onClick = onEndRoute,
                    variant = KozmosButtonVariant.Outline,
                    emotion = KozmosButtonEmotion.Danger,
                    size = KozmosButtonSize.Sm
                ) {
                    Text(endLabel)
                }
            }
            Row(
                verticalAlignment = Alignment.Bottom,
                modifier = Modifier
                    .fillMaxWidth()
                    .semantics(mergeDescendants = true) {}
            ) {
                CompositionLocalProvider(LocalContentColor provides KozmosThemeTokens.primitivesColorsForeground100) {
                    Text(
                        text = durationText,
                        style = MaterialTheme.typography.bodyLarge.copy(fontWeight = FontWeight.SemiBold)
                    )
                    Spacer(modifier = Modifier.size(KozmosDimensions.primitivesLayoutSpacing150))
                    Text(text = distanceText, style = MaterialTheme.typography.bodyLarge)
                    Spacer(modifier = Modifier.weight(1f))
                    if (arrivalText != null) {
                        Text(text = arrivalText, style = MaterialTheme.typography.bodyLarge)
                    }
                }
            }
            progress?.invoke()
        }
    }
}

File: packages/ios/Sources/Components/RouteSummary/RouteSummary.swift
import SwiftUI

public enum KozmosRouteSummaryState {
    case active
    case preview
}

public struct KozmosRouteSummary<TransportModeIcon: View>: View {
    private let etaText: String
    private let distanceText: String
    private let state: KozmosRouteSummaryState
    private let onEndRoute: () -> Void
    private let onStartNavigation: (() -> Void)?
    private let transportModeIcon: TransportModeIcon
    private let showsTransportModeIcon: Bool
    // The navigation layout, when a destination is given: its name with End
    // beside it, the time, distance and arrival on one row, and the caller's
    // progress below. Mirrors the product prototype's navigation sheet.
    private let destination: String?
    private let durationText: String?
    private let arrivalText: String?
    private let endLabel: String
    private let progress: AnyView?

    public init(
        etaText: String,
        distanceText: String,
        state: KozmosRouteSummaryState = .active,
        onEndRoute: @escaping () -> Void,
        onStartNavigation: (() -> Void)? = nil,
        @ViewBuilder transportModeIcon: () -> TransportModeIcon
    ) {
        self.etaText = etaText
        self.distanceText = distanceText
        self.state = state
        self.onEndRoute = onEndRoute
        self.onStartNavigation = onStartNavigation
        self.transportModeIcon = transportModeIcon()
        self.showsTransportModeIcon = true
        self.destination = nil
        self.durationText = nil
        self.arrivalText = nil
        self.endLabel = "End"
        self.progress = nil
    }

    /// The navigation layout: the destination's name with End beside it in
    /// the danger outline; `durationText`, `distanceText` and `arrivalText`
    /// on one row; `progress` — a `KozmosRouteProgressRail` in the products —
    /// below.
    public init(
        destination: String,
        durationText: String,
        distanceText: String,
        arrivalText: String? = nil,
        endLabel: String = "End",
        onEndRoute: @escaping () -> Void,
        @ViewBuilder progress: () -> some View
    ) where TransportModeIcon == EmptyView {
        self.etaText = durationText
        self.distanceText = distanceText
        self.state = .active
        self.onEndRoute = onEndRoute
        self.onStartNavigation = nil
        self.transportModeIcon = EmptyView()
        self.showsTransportModeIcon = false
        self.destination = destination
        self.durationText = durationText
        self.arrivalText = arrivalText
        self.endLabel = endLabel
        self.progress = AnyView(progress())
    }

    public var body: some View {
        if let destination {
            navigation(destination: destination)
        } else {
            summary
        }
    }

    private func navigation(destination: String) -> some View {
        VStack(spacing: KozmosDimensions.primitivesLayoutSpacing150) {
            HStack(alignment: .center, spacing: KozmosDimensions.primitivesLayoutSpacing150) {
                Text(destination)
                    .font(KozmosTypography.title3.weight(.semibold))
                    .foregroundColor(KozmosColors.primitivesColorsForeground100)
                    .lineLimit(2)
                    .fixedSize(horizontal: false, vertical: true)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .accessibilityAddTraits(.isHeader)
                KozmosButton(endLabel, variant: .outline, emotion: .danger, size: .sm, action: onEndRoute)
            }
            HStack(alignment: .firstTextBaseline, spacing: KozmosDimensions.primitivesLayoutSpacing150) {
                Text(durationText ?? etaText)
                    .font(KozmosTypography.subheadline.weight(.semibold))
                    .foregroundColor(KozmosColors.primitivesColorsForeground100)
                Text(distanceText)
                    .font(KozmosTypography.subheadline)
                    .foregroundColor(KozmosColors.primitivesColorsForeground100)
                Spacer(minLength: KozmosDimensions.primitivesLayoutSpacing100)
                if let arrivalText {
                    Text(arrivalText)
                        .font(KozmosTypography.subheadline)
                        .foregroundColor(KozmosColors.primitivesColorsForeground100)
                }
            }
            .accessibilityElement(children: .combine)
            if let progress {
                progress
            }
        }
        .padding(KozmosDimensions.primitivesLayoutSpacing200)
        .kozmosGlassSurface(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusPanel, style: .continuous))
        .kozmosElevation(KozmosShadows.semanticsElevationOverlay)
    }

    private var summary: some View {
        VStack(spacing: KozmosDimensions.primitivesLayoutSpacing200) {
            HStack(spacing: KozmosDimensions.primitivesLayoutSpacing150) {
                if showsTransportModeIcon {
                    transportModeIcon
                        .frame(width: 40, height: 40)
                        .foregroundColor(KozmosColors.primitivesColorsTheme500)
                        .background(KozmosColors.primitivesColorsTheme500.opacity(0.12))
                        .clipShape(Circle())
                        // The words beside it carry the meaning.
                        .accessibilityHidden(true)
                }

                VStack(alignment: .leading, spacing: KozmosDimensions.primitivesLayoutSpacing25) {
                    Text(etaText)
                        .font(.title3.weight(.bold))
                        .foregroundColor(KozmosColors.primitivesColorsForeground100)

                    Text(distanceText)
                        .font(.subheadline.weight(.medium))
                        .foregroundColor(KozmosColors.primitivesColorsForeground500)
                }

                Spacer()

                if state == .active {
                    Button(action: onEndRoute) {
                        Image(systemName: "xmark")
                            .font(.system(size: 16, weight: .bold))
                            .frame(width: 40, height: 40)
                            .foregroundColor(KozmosColors.componentsPrimaryButtonsDangerButtonForegroundContentIdle)
                            .background(KozmosColors.componentsPrimaryButtonsDangerButtonBackgroundIdle)
                            .clipShape(Circle())
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel("End route")
                }
            }

            if state == .preview, let onStartNavigation {
                KozmosButton("Start Navigation", size: .lg, action: onStartNavigation)
                    .frame(maxWidth: .infinity)
            }
        }
        .padding(KozmosDimensions.primitivesLayoutSpacing200)
        .kozmosGlassSurface(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusPanel, style: .continuous))
        .kozmosElevation(KozmosShadows.semanticsElevationOverlay)
    }
}

public extension KozmosRouteSummary where TransportModeIcon == EmptyView {
    init(
        etaText: String,
        distanceText: String,
        state: KozmosRouteSummaryState = .active,
        onEndRoute: @escaping () -> Void,
        onStartNavigation: (() -> Void)? = nil
    ) {
        self.etaText = etaText
        self.distanceText = distanceText
        self.state = state
        self.onEndRoute = onEndRoute
        self.onStartNavigation = onStartNavigation
        self.transportModeIcon = EmptyView()
        self.showsTransportModeIcon = false
        self.destination = nil
        self.durationText = nil
        self.arrivalText = nil
        self.endLabel = "End"
        self.progress = nil
    }
}
